File: antilles-core/openHPC_web_project/antilles/cluster/views/node.py
# -*- coding: utf-8 -*-

"""
Copyright © 2019-present Lenovo

This file is licensed under both the BSD-3 license for individual use and
EPL-1.0 license for commercial use. Full text of both licenses can be found in
COPYING.BSD and COPYING.EPL files.
"""

import json
import logging

from django.db.models import Max, Q
from rest_framework.response import Response
from rest_framework.views import APIView
from six import raise_from

from antilles.alarm.models import Alarm, Policy
from antilles.cluster.exceptions import (
    CheckPreferenceException, PowerOperationException,
)
from antilles.cluster.managers import power
from antilles.cluster.models import Node
from antilles.common.exceptions import InvalidParameterException
from antilles.common.utils import json_schema_validate
from antilles.optlog.optlog import EventLog
from antilles.scheduler.models import RunningJob
from antilles.user.models import Preference
from antilles.user.permissions import AsOperatorRole, AsUserRole

logger = logging.getLogger(__name__)


IDLE_THRESHOLD = 20
BUSY_THRESHOLD = 80


class NodeBase(object):
    columns_mapping = {}

    def get_range(self, *args, **kwargs):
        return Node.objects.all()

    def get_node_status_preference(self, user):
        logger.info("Get preference for node status")
        name = "monitor.policy.node.status"
        try:
            preference = Preference.objects.get(name=name)
        except Preference.DoesNotExist as e:
            logger.exception('Preference name %s does not exist', name)
            raise_from(CheckPreferenceException, e)
        else:
            if preference.user in [None, user]:
                # 'cpu_core' or 'cpu_util' for value
                return preference.value
            else:
                return "cpu_util"

    def convert_cpu_status(self, preference, cpu_util, node_id):
        if preference == "cpu_util":
            util = int(cpu_util)
            if util < IDLE_THRESHOLD:
                return "idle"
            elif util > BUSY_THRESHOLD:
                return "busy"
            else:
                return "used"
        else:
            if RunningJob.objects.filter(node__id=node_id).count() > 0:
                return "busy"
            else:
                return "idle"

    def search(self, query, argss):
        if 'search' not in argss:
            return query
        else:
            search = argss['search']
            props = search['props']
            keyword = search['keyword'].strip()
            if keyword == "":
                return query

            q = Q()
            for field in props:
                prop = field
                if field in self.columns_mapping:
                    prop = self.columns_mapping[field]
                q |= Q(**{prop + "__icontains": keyword})

            return query.filter(q)

    def sort_fields(self, argss):
        if 'sort' not in argss:
            return ['id'], True
        order = argss["sort"]
        prop = order['prop']
        if prop in self.columns_mapping:
            prop = self.columns_mapping[order['prop']]
        if order['order'] != "ascending":
            prop = "-" + prop
        return prop

    def sort(self, query, argss):
        return query.order_by(self.sort_fields(argss))

    def filters(self, query, filters):

        if len(filters) > 0 and filters is not None:
            for field in filters:
                if field['prop'] in self.columns_mapping:
                    prop = self.columns_mapping[field['prop']]
                else:
                    prop = field['prop']

                if len(field["values"]) > 0:
                    key = prop + "__{}".format(field["type"])
                    query = query.filter(
                        **{
                            key: field["values"]
                        }
                    )
        return query

    def params(self, args):
        return args


class NodeList(NodeBase, APIView):
    permission_classes = (AsUserRole,)

    columns_mapping = {
        "status": "cpu_util",
        "groups": "groups__name"
    }

    def get(self, request):
        argss = json.loads(
            request.GET["args"]
        )

        argss = self.params(argss)

        query = self.get_range()

        query = self.filters(query, argss['filters'])

        query = self.search(query, argss)

        query = self.sort(query, argss)

        filtered_total = query.count()
        offset = argss['offset'] \
            if int(argss['offset']) < filtered_total else 0
        results = query[offset:offset + int(argss['length'])]
        offset = offset + len(results)

        preference = self.get_node_status_preference(request.user)

        return Response(
            {
                'offset': offset,
                'total': filtered_total,
                'data': [self.trans_result(result, preference)
                         for result in results],
            }
        )

    def trans_result(self, result, preference):
        groups = []
        for grp in result.groups.iterator():
            groups.append(grp.name)

        if result.gpu:
            gpus = result.gpu.all()
            gpu_cnt = gpus.count()
            gpu_used = [0] * gpu_cnt
            gpu_type = [""] * gpu_cnt
            for gpu in gpus:
                gpu_used[gpu.index] = int(gpu.occupation)
                gpu_type[gpu.index] = gpu.type
        else:
            gpu_used = []
            gpu_type = []

        power_status = status = "off"
        if result.power_status is True:
            power_status = "on"
            status = self.convert_cpu_status(
                preference,
                result.cpu_util,
                result.id
            )

        return {
            "id": result.id,
            "hostname": result.hostname,
            "bmc_ipv4": result.bmc_ipv4,
            "disk_total": int(result.disk_total),
            "memory_total": int(result.memory_total),
            "mgt_ipv4": result.mgt_ipv4,
            "type": result.type,
            "processors_total": result.cpu_total,
            "power_status": power_status,
            "groups": ",".join(groups),
            "status": status,
            "gpus": {
                "used": gpu_used,
                "type": gpu_type
            }
        }


class NodeAll(APIView):
    columns_mapping = {
        'all': "all",
        'login': "login",
        'head': "head",
        'compute': "compute",
        'io': "io",
    }

    def get(self, request):
        if 'type' in request.GET:
            node_type = request.GET['type'].strip()
            node_type = self.columns_mapping[node_type.lower()] \
                if node_type.lower() in self.columns_mapping else ""
        else:
            raise InvalidParameterException
        query = Node.objects.iterator() if 'all' == node_type else\
            Node.objects.filter(type=node_type).iterator()
        return Response(
            {
                'nodelist': [
                    {'id': q.id, 'name': q.hostname} for q in query
                ]
            }
        )


class NodeDetail(NodeBase, APIView):
    permission_classes = (AsOperatorRole,)

    def get_range(self, pk):
        return Node.objects.get(id=pk)

    def get(self, request, pk, format=None):
        query = self.get_range(pk)

        groups = []
        for grp in query.groups.iterator():
            groups.append(grp.name)

        if query.gpu:
            gpus = query.gpu.all()
            gpu_cnt = gpus.count()
            gpu_used = [0] * gpu_cnt
            gpu_type = [""] * gpu_cnt
            for gpu in gpus:
                gpu_used[gpu.index] = int(gpu.occupation)
                gpu_type[gpu.index] = gpu.type
        else:
            gpu_used = []
            gpu_type = []

        preference = self.get_node_status_preference(request.user)
        status = "off"
        power_status = "off"
        if query.power_status is True:
            power_status = "on"
            status = self.convert_cpu_status(
                preference,
                query.cpu_util,
                query.id
            )

        alarm_level = Alarm.objects \
            .exclude(status=Alarm.RESOLVED) \
            .filter(node=query.hostname) \
            .aggregate(level=Max('policy__level')) \
            .get('level', Policy.NOTSET)

        return Response({
            "node": {
                "alarm_level": alarm_level,
                "bmc_ipv4": query.bmc_ipv4,
                "disk_total": int(query.disk_total),
                "disk_used": int(query.disk_used),
                "hostname": query.hostname,
                "id": query.id,
                "machinetype": query.machinetype,
                "memory_total": int(query.memory_total),
                "memory_used": int(query.memory_used),
                "mgt_ipv4": query.mgt_ipv4,
                "power_status": power_status,
                "processors_total": query.cpu_total,
                "status": status,
                "type": query.type,
                "groups": ",".join(groups),
                "gpus": {
                    "used": gpu_used,
                    "type": gpu_type
                }
            }
        })

    @json_schema_validate({
        'type': 'object',
        'properties': {
            'operation': {
                'type': 'string',
                'enum': ['turn_on', 'turn_off']
            },
            'bootmode': {
                'type': 'string'
            },
            'nextdevice': {
                'type': 'string'
            },
            'persistent': {
                'type': 'boolean'
            },
        },
        'required': [
            'operation'
        ]
    })
    def put(self, request, pk):
        node = Node.objects.get(id=pk)
        hostname = node.hostname
        if 'operation' in request.data:
            operation = request.data['operation']
        else:
            raise InvalidParameterException

        try:
            if operation == 'turn_off':
                power.shutdown_device(node)
                node.power_status = False
                node.save()
            else:
                power.startup_device(
                    node,
                    request.data.get('bootmode', 'uefi'),
                    request.data.get('nextdevice', None),
                    request.data.get('persistent', 'False')
                )
                node.power_status = True
                node.save()

        except Exception:
            raise PowerOperationException
        else:
            # add operationlog
            EventLog.opt_create(request.user.username, EventLog.node,
                                getattr(EventLog, operation),
                                EventLog.make_list(pk, hostname)
                                )

        return Response()

File: antilles-portal/src/service/node.js
/*
 * Copyright © 2019-present Lenovo
 * 
 * This file is licensed under both the BSD-3 license for individual/non-commercial use and
 * EPL-1.0 license for commercial use. Full text of both licenses can be found in
 * COPYING.BSD and COPYING.EPL files.
 */

import Vue from 'vue'
import TableDataFetcherFactory from '../common/table-data-fetcher-factory'
import ErrorHandler from '../common/error-handler'
import AlarmPolicyService from './alarm-policy'
import GPUService from './gpu'

const NodeStatusEnums = ['off', 'idle', 'used', 'busy'];
const NodePowerStatusEnums = ['on', 'off'];
const NodeType = ['head', 'login', 'compute', 'io'];

class NodeLocation {
  constructor() {
    this.slot = 0;
    this.height = 0;
    this.width = 0;
    this.rackId = 0;
    this.chassisId = 0;
  }
  static parseFromRestApi(jsonObj) {
    let location = new NodeLocation();
    location.slot = jsonObj.u;
    location.height = parseInt(jsonObj.height);
    location.width = parseFloat(jsonObj.width);
    location.rackId = jsonObj.rack_id;
    location.chassisId = jsonObj.chassis_id;
  }
  get slot() {
    return this._slot;
  }
  set slot(slot) {
   this._slot = slot;
  }
  get height() {
    return this._height;
  }
  set height(height) {
   this._height = height;
  }
  get width() {
    return this._width;
  }
  set width(width) {
   this._width = width;
  }
  get rackId() {
    return this._rackId;
  }
  set rackId(rackId) {
   this._rackId = rackId;
  }
  get chassisId() {
    return this._chassisId;
  }
  set chassisId(chassisId) {
   this._chassisId = chassisId;
  }
}

class Gpu {
  constructor() {
    this.index = 0;
    this.type = '';
    this.used = false;
  }
  get index() {
    return this._index;
  }
  set index(index) {
    this._index = index;
  }
  get type() {
    return this._type;
  }
  set type(type) {
    this._type = type;
  }
  get used() {
    return this._used;
  }
  set used(used) {
    this._used = used;
  }
}

class Node {
  constructor() {
    this.id = 0;
    this.hostname = '';
    this.status = '';
    this.powerStatus = '';
    this.type = '';
    this.osIP = '';
    this.bmcIP = '';
    this.machineType = '';
    this.frontImageUrl = '';
    this.groups = [];
    this.cpuUsed = 0.0;
    this.cpuTotal = 0;
    this.ramUsed = 0;
    this.ramTotal = 0;
    this.diskUsed = 0;
    this.diskTotal = 0;
    this.alarmPolicyLevel = '';
    this.location = new NodeLocation();
    this.gpus = [];
  }
  static parseFromRestApi(jsonObj){
    let node = new Node();
    node.id = jsonObj.id;
    node.hostname = jsonObj.hostname;
    node.status = jsonObj.status.toLowerCase()=='used'?'busy':jsonObj.status.toLowerCase();
    node.powerStatus = jsonObj.power_status.toLowerCase();
    node.type = jsonObj.type.toLowerCase();
    node.osIP = jsonObj.mgt_ipv4;
    node.bmcIP = jsonObj.bmc_ipv4;
    node.machineType = jsonObj.machinetype;
    node.frontImageUrl = jsonObj.frontimage;
    node.groups = jsonObj.groups;
    node.cpuUsed = 0.0;
    node.cpuTotal = jsonObj.processors_total;
    node.ramUsed = jsonObj.memory_used * 1024;
    node.ramTotal = jsonObj.memory_total * 1024;
    node.diskUsed = jsonObj.disk_used * 1024 * 1024 * 1024;
    node.diskTotal = jsonObj.disk_total * 1024 * 1024 * 1024;
    node.alarmPolicyLevel = jsonObj.alarm_level == null ? null : AlarmPolicyService.AlarmLevelToParse[String(jsonObj.alarm_level)];
    node.location = jsonObj.location?NodeLocation.parseFromRestApi(jsonObj.location):'';
    node.gpus = [];
    if(jsonObj.gpus) {
      for(var i=0; i<jsonObj.gpus.type.length; i++) {
        var gpu = new Gpu();
        gpu.index = i;
        gpu.type = jsonObj.gpus.type[i];
        gpu.used = jsonObj.gpus.used[i] == 1 ? true : false;
        node.gpus.push(gpu);
      }
    }
    return node;
  }
  get id() {
    return this._id;
  }
  set id(id) {
   this._id = id;
  }
  get hostname() {
    return this._hostname;
  }
  set hostname(hostname) {
   this._hostname = hostname;
  }
  get status() {
    return this._status;
  }
  set status(status) {
   this._status = status;
  }
  get powerStatus() {
    return this._powerStatus;
  }
  set powerStatus(powerStatus) {
   this._powerStatus = powerStatus;
  }
  get type() {
    return this._type;
  }
  set type(type) {
   this._type = type;
  }
  get osIP() {
    return this._osIP;
  }
  set osIP(osIP) {
   this._osIP = osIP;
  }
  get bmcIP() {
    return this._bmcIP;
  }
  set bmcIP(bmcIP) {
   this._bmcIP = bmcIP;
  }
  get machineType() {
    return this._machineType;
  }
  set machineType(machineType) {
   this._machineType = machineType;
  }
  get frontImageUrl() {
    return this._frontImageUrl;
  }
  set frontImageUrl(frontImageUrl) {
   this._frontImageUrl = frontImageUrl;
  }
  get groups() {
    return this._groups;
  }
  set groups(groups) {
   this._groups = groups;
  }
  get cpuUsed() {
    return this._cpuUsed;
  }
  set cpuUsed(cpuUsed) {
   this._cpuUsed = cpuUsed;
  }
  get cpuTotal() {
    return this._cpuTotal;
  }
  set cpuTotal(cpuTotal) {
   this._cpuTotal = cpuTotal;
  }
  get ramUsed() {
    return this._ramUsed;
  }
  set ramUsed(ramUsed) {
   this._ramUsed = ramUsed;
  }
  get ramTotal() {
    return this._ramTotal;
  }
  set ramTotal(ramTotal) {
   this._ramTotal = ramTotal;
  }
  get diskUsed() {
    return this._diskUsed;
  }
  set diskUsed(diskUsed) {
   this._diskUsed = diskUsed;
  }
  get diskTotal() {
    return this._diskTotal;
  }
  set diskTotal(diskTotal) {
   this._diskTotal = diskTotal;
  }
  get alarmPolicyLevel() {
    return this._alarmPolicyLevel;
  }
  set alarmPolicyLevel(alarmPolicyLevel) {
   this._alarmPolicyLevel = alarmPolicyLevel;
  }
  get location() {
    return this._location;
  }
  set location(location) {
   this._location = location;
  }
  get gpuUsed() {
    return this._gpuUsed;
  }
  set gpuUsed(gpuUsed) {
   this._gpuUsed = gpuUsed;
  }
  get gpuTotal() {
    return this._gpuTotal;
  }
  set gpuTotal(gpuTotal) {
   this._gpuTotal = gpuTotal;
  }
  get gpuModel() {
    return this._gpuModel;
  }
  set gpuModel(gpuModel) {
    this._gpuModel = gpuModel;
  }
}

function nodesTableDataParser(res) {
  var nodes = [];
  res.data.forEach((item) => {
    nodes.push(Node.parseFromRestApi(item));
  });
  return {
    offset: res.offset,
    total: res.total,
    data: nodes
  };
}

function nodesRestApiPropMap(prop) {
  if(prop=='powerStatus')
    return 'power_status';
  if(prop=='osIP')
    return 'mgt_ipv4';
  if(prop=='bmcIP')
    return 'bmc_ipv4';
  if(prop=='machineType')
    return 'machinetype';
  if(prop=='cpuTotal')
    return 'processors_total';
  if(prop=='ramUsed')
    return 'memory_used';
  if(prop=='ramTotal')
    return 'memory_total';
  if(prop=='diskUsed')
    return 'disk_used';
  if(prop=='diskTotal')
    return 'disk_total';
  if(prop=='alarmPolicyLevel')
    return 'alarm_level';
  return prop;
}

function getNodesTableDataFetcher() {
  return TableDataFetcherFactory.createRemotePagingFetcher('/api/nodes/', nodesRestApiPropMap, nodesTableDataParser, 'data', 'offset', 'total');
}

function powerOnNode(nodeId, nextDevice) {
  var req = {
    operation: 'turn_on'
  };
  if(nextDevice) {
    req.bootmode = 'uefi';
    req.nextdevice = nextDevice;
    req.persistent = 'False';
  }
  return new Promise((resolve, reject) => {
    Vue.http.put('/api/nodes/'+ nodeId + '/', req).then((res) => {
        resolve(nodeId);
      },
      (res) => {
        ErrorHandler.restApiErrorHandler(res, reject);
      }
    );
  });
}

function powerOffNode(nodeId) {
  var req = {
    operation: "turn_off"
  }
  return new Promise((resolve, reject) => {
    Vue.http.put('/api/nodes/'+ nodeId + '/', req).then((res) => {
        resolve(nodeId);
      },
      (res) => {
        ErrorHandler.restApiErrorHandler(res, reject);
      }
    );
  });
}

function getNodeById(id) {
  return new Promise((resolve, reject) => {
    Vue.http.get('/api/nodes/' + id).then((res) => {
      var node = Node.parseFromRestApi(res.body.node);
      resolve(node);
    },(res) => {
      ErrorHandler.restApiErrorHandler(res, reject);
    })
  });
}

function getAllNodes(type) {
  return new Promise((resolve, reject) => {
    Vue.http.get('/api/nodes/nodelist/?type='+type).then((res) => {
      var data = [];
      res.body.nodelist.forEach((node) => {
        data.push({
          hostname: node.name,
          id: node.id
        });
      })
      resolve(data)
    },(res) => {
      ErrorHandler.restApiErrorHandler(res, reject);
    })
  });
}

function getNodeConsoleServiceUrl(hostname) {
  return '/api/nodes/' + hostname + '/console/sessions/';
}

function getNodeShellServiceUrl(hostname) {
  return '/api/nodes/' + hostname + '/shell/sessions/';
}

export default {
  getNodesTableDataFetcher,
  powerOnNode,
  powerOffNode,
  getNodeById,
  getAllNodes,
  NodeStatusEnums,
  NodePowerStatusEnums,
  NodeType,
  getNodeConsoleServiceUrl,
  getNodeShellServiceUrl
}
